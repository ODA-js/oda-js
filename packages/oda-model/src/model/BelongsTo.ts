import { Record } from 'immutable';

import { IBelongsTo, IBelongsToInit, IBelongsToStore, IRelationTransform } from '../interfaces/IBelongsTo';
import { IEntityRef } from '../interfaces/IEntityRef';
import { Relation } from './Relation';
import { TransformField, TransformRef } from './utils';

// tslint:disable-next-line:variable-name
export const DefaultBelongsTo: IBelongsToStore = {
  name: null,
  title: null,
  description: null,
  fields: null,
  opposite: null,
  belongsTo: null,
  //storage
  single: true,
  stored: true,
  embedded: true,
  // name
  fullName: null,
  normalName: null,
  shortName: null,
};

// tslint:disable-next-line:variable-name
export const BelongsToTransform: IRelationTransform = {
  belongsTo: TransformRef(),
  fields: TransformField(),
};

// tslint:disable-next-line:variable-name
export const BelongsToStorage = Record(DefaultBelongsTo);

export class BelongsTo extends Relation<IBelongsToInit, IBelongsToStore> implements IBelongsTo {
  public get verb(): 'BelongsTo' {
    return 'BelongsTo';
  }
  public get ref(): IEntityRef {
    return this.store.get('belongsTo', null);
  }
  public get belongsTo(): IEntityRef {
    return this.store.get('belongsTo', null);
  }

  protected transform(input: Partial<IBelongsToInit>): Partial<IBelongsToStore> {
    const result: Partial<IBelongsToStore> = {};
    if (input) {
      for (let f in input) {
        if (input.hasOwnProperty(f)) {
          if (f === 'belongsTo') {
            result.belongsTo = BelongsToTransform.belongsTo.transform(input.belongsTo);
          } else if (f === 'fields') {
            result.fields = BelongsToTransform.fields.transform(input.fields);
          } else {
            result[f] = input[f];
          }
        }
      }
    }
    return result;
  }

  protected reverse(input: Record<IBelongsToStore> & Readonly<IBelongsToStore>): Partial<IBelongsToInit> {
    const result: Partial<IBelongsToInit> = {};
    if (input) {
      const core = input.toJS();
      for (let f in core) {
        if (core.hasOwnProperty(f)) {
          if (f === 'belongsTo' || f === 'ref') {
            result.belongsTo = BelongsToTransform.belongsTo.reverse(input.belongsTo);
          } else if (f === 'fields') {
            result.fields = BelongsToTransform.fields.reverse(input.fields);
          } else {
            result[f] = core[f];
          }
        }
      }
    }
    return result;
  }

  constructor(init: Partial<IBelongsToInit> = {}) {
    super();
    this.store = new BelongsToStorage(this.transform(init));
    this.init = new (Record<Partial<IBelongsToInit>>(init))();
  }
}
