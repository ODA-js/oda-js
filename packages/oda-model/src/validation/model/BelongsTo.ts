import { IRelationPropsStore } from '../interfaces/IRelation';
import { Record } from 'immutable';
import { Map, Set } from 'immutable';

import { Persistent } from './Persistent';
import { transformMap, transformSet } from './utils';
import { IBelongsToRelationPropsStore, IBelongsToRelationProps, IBelongsToRelation } from '../interfaces/IBelongsToRelation';
import { IEntityRef } from '../interfaces/IEntityRef';
import { IField } from '../interfaces/IField';

// tslint:disable-next-line:variable-name
export const DefaultBelongsTo: IBelongsToRelationPropsStore = {
  modelType: 'relation',
  name: null,
  title: null,
  description: null,
  verb: null,
  ref: null,
  fields: null,
  opposite: null,
  belongsTo: null,
};

// tslint:disable-next-line:variable-name
export const RelationTransform: {[k in keyof IBelongsToRelationPropsStore]?: any } = {
  fields: transformMap<IField>(),
};

// tslint:disable-next-line:variable-name
export const BelongsToStorage = Record(DefaultBelongsTo);

export class BelongsTo extends Persistent<IBelongsToRelationProps, IBelongsToRelationPropsStore> implements IBelongsToRelation {
  public get modelType(): 'relation' {
    return 'relation';
  }
  public get name(): string {
    return this.store.get('name', null);
  }
  public get title(): string {
    return this.store.get('title', null);
  }
  public get description(): string {
    return this.store.get('description', null);
  }
  public get ref(): IEntityRef {
    return this.store.get('ref', null);
  }
  public get fields(): Map<string, IField> {
    return this.store.get('fields', null);
  }
  public get opposite(): string {
    return this.store.get('opposite', null);
  }

  public get verb(): 'BelongsTo' {
    return 'BelongsTo';
  }
  public get belongsTo(): IEntityRef {
    return this.store.get('belongsTo', null);
  }

  protected transform(input: IBelongsToRelationProps): IBelongsToRelationPropsStore {
    return {
      ...input,
      fields: RelationTransform.fields.transform(input.fields),
    };
  }
  protected reverse(input: IBelongsToRelationPropsStore): IBelongsToRelationProps {
    return {
      ...input,
      fields: RelationTransform.fields.reverse(input.fields),
    };
  }
  constructor(init: IBelongsToRelationProps) {
    super();
    this.store = new BelongsToStorage(this.transform(init));
    this.init = new (Record<IBelongsToRelationProps>(init))();
  }
}
