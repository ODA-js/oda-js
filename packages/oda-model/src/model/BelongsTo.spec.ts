import { BelongsTo } from './BelongsTo';

describe('BelongsTo', () => {
  let relation: BelongsTo;

  beforeAll(() => {
    expect(() => new BelongsTo()).not.toThrow();
  });

  beforeEach(() => {
    relation = new BelongsTo();
  });

  it('create empty', () => {
    expect(relation).toMatchSnapshot();
    expect(relation.toJS()).toMatchSnapshot();
  });

  it('updates strings', () => {
    expect(() => relation.updateWith({
      name: 'cool',
      description: 'very cool',
      title: 'very cool title',
    })).not.toThrow();
    expect(relation).toMatchSnapshot();
  });

  it('toJS with dupes', () => {
    expect(() => relation.updateWith({
      fields: [
        { name: 'one' },
        { name: 'one' },
      ],
    },
    )).not.toThrow();

    expect(relation.ref).toBeNull();
    expect(relation.fields.size).toBe(1);
    expect(relation.toJS()).toMatchSnapshot();
  });

  it('toJS with HashMap', () => {
    expect(() => relation.updateWith({
      fields: {
        one: {},
      },
    },
    )).not.toThrow();

    expect(relation.ref).toBeNull();
    expect(relation.fields.size).toBe(1);
    expect(relation.toJS()).toMatchObject({
      fields: [
        { name: 'one' },
      ],
    });
  });

  it('toJS ', () => {
    expect(() => relation.updateWith({
      belongsTo: 'id@Some#id',
    },
    )).not.toThrow();

    expect(relation.ref).not.toBeNull();
    expect(relation.ref.backField).toBe('id');
    expect(relation.ref.entity).toBe('Some');
    expect(relation.ref.field).toBe('id');

    expect(relation.ref.toJS()).toMatchObject({
      entity: 'Some',
      backField: 'id',
      field: 'id',
    });

    expect(relation.toJS()).toMatchObject({
      belongsTo: 'id@Some#id',
    });
  });
});
