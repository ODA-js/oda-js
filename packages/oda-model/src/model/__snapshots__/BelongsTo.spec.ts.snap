// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BelongsTo create empty 1`] = `
BelongsTo {
  "init": Immutable.Record {},
  "store": Immutable.Record {
    "name": null,
    "title": null,
    "description": null,
    "fields": null,
    "opposite": null,
    "belongsTo": null,
    "single": true,
    "stored": true,
    "embedded": true,
    "fullName": null,
    "normalName": null,
    "shortName": null,
  },
}
`;

exports[`BelongsTo create empty 2`] = `
Object {
  "belongsTo": null,
  "description": null,
  "embedded": true,
  "fields": null,
  "fullName": null,
  "name": null,
  "normalName": null,
  "opposite": null,
  "shortName": null,
  "single": true,
  "stored": true,
  "title": null,
}
`;

exports[`BelongsTo toJS with dupes 1`] = `
Object {
  "belongsTo": null,
  "description": null,
  "embedded": true,
  "fields": Array [
    Object {
      "acl": null,
      "args": null,
      "derived": null,
      "description": null,
      "entity": null,
      "idKey": null,
      "identity": null,
      "indexed": null,
      "name": "one",
      "order": null,
      "persistent": null,
      "relation": null,
      "required": null,
      "title": null,
      "type": null,
    },
  ],
  "fullName": null,
  "name": null,
  "normalName": null,
  "opposite": null,
  "shortName": null,
  "single": true,
  "stored": true,
  "title": null,
}
`;

exports[`BelongsTo updates strings 1`] = `
BelongsTo {
  "init": Immutable.Record {},
  "store": Immutable.Record {
    "name": "cool",
    "title": "very cool title",
    "description": "very cool",
    "fields": null,
    "opposite": null,
    "belongsTo": null,
    "single": true,
    "stored": true,
    "embedded": true,
    "fullName": null,
    "normalName": null,
    "shortName": null,
  },
}
`;
